examples/with_plugs.py

```diff
@@ -0,0 +1,114 @@
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Example OpenHTF test logic.
+
+Run with (your virtualenv must be activated first):
+
+  python with_plugs.py
+
+Afterwards, take a look at the hello_world.json output file.  This will
+give you a basic idea of what a minimal test outputs.
+
+For more information on measurements, see the measurements.py example.
+
+TODO(someone): Write an output example.
+For more information on output, see the output.py example.
+
+with_plugs() is most useful when you have a test that has to happen in the same
+(or mostly similar) ways across multiple interfaces, such testing connectivity
+on a quad-NIC board. Instead of creating 4 phases with 4 plugs, you create 1 phase
+with 4 subclasses of the same plug and use with_plugs() to create the 4 phases.
+"""
+import itertools
+import subprocess
+import time
+
+import openhtf as htf
+from openhtf import plugs
+
+
+class PingPlug(plugs.BasePlug):
+  """This plug simply does a ping against the host attribute."""
+  enable_remote = True
+  host = None
+
+  def __init__(self):
+    # This should only ever be constructed by a subclass, so this is only
+    # checking that the subclass set host correctly.
+    assert self.host is not None
+
+  def _get_command(self, count):
+    # Returns the commandline for pinging the host.
+    return [
+      'ping',
+      '-c',
+      str(count),
+      self.host,
+    ]
+
+  def run(self, count):
+    command = self._get_command(count)
+    print "running: %s" % command
+    return subprocess.call(command)
+
+
+# These subclasses specify the host to ping so they can be used directly by
+# phases or through with_plugs().
+class PingGoogle(PingPlug):
+  host = 'google.com'
+
+
+class PingDnsA(PingPlug):
+  host = '8.8.8.8'
+
+
+class PingDnsB(PingPlug):
+  host = '8.8.4.4'
+
+
+@plugs.plug(pinger=PingPlug.placeholder)
+@htf.measures('total_time', 'retcode')
+def test_ping(test, pinger, count):
+  """This tests that we can ping a host.
+
+  The plug, pinger, is expected to be replaced at test creation time, so the
+  placeholder property was used instead of the class directly.
+  """
+  start = time.time()
+  retcode = pinger.run(count)
+  elapsed = time.time() - start
+  test.measurements.total_time = elapsed
+  test.measurements.retcode = retcode
+
+
+if __name__ == '__main__':
+  # We instantiate our OpenHTF test with the phases we want to run as args.
+
+  # We're going to use these counts and these plugs to create all our phases
+  # using only 1 written phase.
+  counts = [3, 5, 10]
+  ping_plugs = [
+    PingGoogle,
+    PingDnsA,
+    PingDnsB,
+  ]
+
+  test = htf.Test(
+      *(test_ping.with_args(count=count).with_plugs(pinger=plug)
+        for count, plug in itertools.product(counts, ping_plugs)))
```

Comment:
no need to use things like `itertools.product` in an example.  we're not trying to teach people fancy python, the target audience should need to know little more than basic python syntax.  the use of `*generator` is pushing the level of complexity for an example, but since that's almost certainly how one would use `.with_plugs()`, it's reasonable to include.